### ISSUE TITLE
Obsolete ti.cfg.X

### ISSUE THREAD

**Issue by yuanming-hu** [taichi-dev/taichi#455]

**Comment by yuanming-hu** [comment]
1. Remove all usages of `ti.cfg` in the codebase (Please see https://github.com/taichi-dev/taichi/blob/master/misc/changelog.md and search for `ti.cfg` to learn about what happened to it.)
2. Delete `ti.cfg` definition

**Comment by archibate** [comment]
> Delete ti.cfg definition

I refuse unless we have some other way to change `CompilerConfig` at runtime, for example:
```py
ti.cfg.some_compiler_config_var = True
kern1()
ti.cfg.some_compiler_config_var = False
kern2()
```
Or, you mean that `CompilerConfig` is **expected to be constant** per `ti.init()`?

**Comment by yuanming-hu** [comment]
> > Delete ti.cfg definition
> 
> I refuse unless we have some other way to change `CompilerConfig` at runtime, for example:
> 
> ```python
> ti.cfg.some_compiler_config_var = True
> kern1()
> ti.cfg.some_compiler_config_var = False
> kern2()
> ```
> 
> Or, you mean that `CompilerConfig` is **expected to be constant** per `ti.init()`?

Good point - we want `CompilerConfig` to be constant for now. We should add a feature to let users specify configs for each kernel (probably using a decorator) later.

**Comment by archibate** [comment]
> probably using a decorator

Then the compiler_config should be a member of Kernel, instead of Program?
So we copy the cfg in Program to Kernel once it's created?

**Comment by yuanming-hu** [comment]
That's one possible solution. For now, let's simply remove `ti.cfg`.

**Comment by TH3CHARLie** [comment]
I take a quick look at this issue, for assignments: `ti.cfg.xxx = yyy`, they can be replaced with `ti.init` according to the changelog. However, I am not sure about accessing (e.g: `if ti.cfg.arch == ti.opengl`), is there a strategy to replace it? I did some search through the codebase for hints but no luck.

**Comment by yuanming-hu** [comment]
I realized that we should probably remove `ti.default_cfg()` as well. Every time we access `ti.cfg`, it should be the function `ti.current_cfg()`, after the `Program` is created.

In `ti.init`, instead of mutating global states, we should probably create a local `CompileConfig` (add a binding via python11 if necessary) and pass it to the initializer of `Program`.

**Comment by TH3CHARLie** [comment]
> I realized that we should probably remove `ti.default_cfg()` as well. Every time we access `ti.cfg`, it should be the function `ti.current_cfg()`, after the `Program` is created.
> 
> In `ti.init`, instead of mutating global states, we should probably create a local `CompileConfig` (add a binding via python11 if necessary) and pass it to the initializer of `Program`.

Thanks for the clarification. For the `ti.init` modification it seems straightforward to me. But I am not sure I get you on `ti.cfg`:

- Are we removing `ti.default_cfg()` usages from impl.py and replace them with `ti.current_cfg()`?

- I am still not sure how we can access the config attributes(for example: `ti.cfg.arch`) if we are removing `ti.cfg` from test/example scripts.

**Comment by yuanming-hu** [comment]
Good questions.



> * Are we removing `ti.default_cfg()` usages from impl.py and replace them with `ti.current_cfg()`?

In `set_default_fp` and `set_default_ip`, we should remove `default_cfg()`.

In `ti.init`, we create a `CompileConfig` instance, set default_fp/ip there. Note that we still need to set the default fp/ip of PyTaichi.



> * I am still not sure how we can access the config attributes(for example: `ti.cfg.arch`) if we are removing `ti.cfg` from test/example scripts.

Let's keep `ti.current_cfg()` for now. Note that this should only be called after `ti.init()`.
